### PR TITLE
[skip_changelog] Form login disabled after user uses OAuth login method

### PR DESCRIPTION
**Pull request by abhvsn**
## Description

With this change, we are implementing a hierarchical login flow where OAuth takes precedence over Form login. So in a nutshell users will have a single login method at any point in time. 

Scenario 1: User signs up with FORM method, user can use form login until she uses OAuth login after that only OAuth method will be enabled
Scenario 2: User signs up with OAuth then only the OAuth method is available. User can't sign in with the form method  

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: oauth-signin-config 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 53.3 **(0)** | 34.81 **(-0.01)** | 31.39 **(0)** | 53.85 **(0)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 50.47 **(-0.24)** | 37.72 **(-0.88)** | 36.21 **(0)** | 55.08 **(-0.27)**</details>

### CODE REVIEW

**Comment by sharat87** [comment]
@abhvsn, can you provide some detail on what the change is please? Like, is it fixing a bug where form login is disabled if the user uses OAuth login method? Or are we disabling form login for users using OAuth? 🤔

**Comment by abhvsn** [comment]
@sharat87 Sorry for not being clear earlier. Here we are not allowing the form login once the user uses the OAuth login method

File: app/server/appsmith-server/src/main/java/com/appsmith/server/authentication/handlers/CustomFormLoginServiceImpl.java [hunk 1]
```diff
@@ -36,6 +38,19 @@ public CustomFormLoginServiceImpl(UserRepository repository) {
...
+                .map(user -> {
+                    // As this will be used for form login only, if the password field is null we can assume this login
+                    // request will fail because of invalid credentials but actual reason might be user has signed up
+                    // using form and then used OAuth for login which removes the password field from user object for
+                    // more details refer AuthenticationSuccessHandler
+                    if (user.getPassword() == null && !LoginSource.FORM.equals(user.getSource())) {
+                        // We can have a implementation to give which login method user should use but this will
+                        // expose the sign-in source for external world and in turn to spammers
+                        throw new InternalAuthenticationServiceException(
+                            "Please use " + user.getSource().toString() + " authentication as login service"
```
Comment (mohanarpit) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/authentication/handlers/CustomFormLoginServiceImpl.java:50]: Nit: Can we use the string message from a `AppsmithError` enum? This ensures that user visible errors are listed in a single place.

**Comment by sharat87** [comment]
I see that the opposite is not being restricted here. For someone who initially signed up with the form, we are allowing them to use OAuth. But once they use OAuth, there's no going back. Their password is cleared and they can no longer use the form to sign in. Is this intended?

**Comment by abhvsn** [comment]
@sharat87 Yes this is expected behavior. Any other way to implement this considering we don't wat to do email verification on sign-up 🤔

**Review by rishabhsaxena — DISMISSED** [review]

File: app/server/appsmith-server/src/main/java/com/appsmith/server/exceptions/AppsmithError.java [hunk 2]
```diff
@@ -62,6 +62,7 @@
     GENERIC_BAD_REQUEST(400, 4028, "Bad Request: {0}", AppsmithErrorAction.DEFAULT, null),
     VALIDATION_FAILURE(400, 4028, "Validation Failure(s): {0}", AppsmithErrorAction.DEFAULT, null),
     INVALID_CURL_COMMAND(400, 4029, "Invalid cURL command, couldn't import.", AppsmithErrorAction.DEFAULT, null),
+    INVALID_LOGIN_METHOD(401, 4030, "Please use {0} authentication as login service.", AppsmithErrorAction.DEFAULT, null),
```
Comment (mohanarpit) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/exceptions/AppsmithError.java:65]: Nit: 
```suggestion
    INVALID_LOGIN_METHOD(401, 4030, "Please use {0} authentication to login to Appsmith", AppsmithErrorAction.DEFAULT, null),
```

**Review by mohanarpit — APPROVED** [review]

**Pull request merged**